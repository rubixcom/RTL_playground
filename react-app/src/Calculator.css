.App {
  text-align: center;
  background: rgb(4, 28, 67);
  height: 100vh;
}

.Calculator {
  display: grid;
  max-width: 700px;
  grid-template-columns: 25% 25% 25% 25%;
  grid-template-rows: auto;
  background: rgb(4, 28, 67);
  grid-template-areas: 
    "ButtonArea ButtonArea ButtonArea ButtonArea";
}

.DisplayPanel {
  grid-area: ButtonArea;
  margin: 10px;
  padding: 10px;
  border-radius: 10px;
  font-size: 50px;
  font-weight: 800;
  text-align: right;
  color: white;
  background: rgb(4, 28, 67);
  border: 3px #3079ED solid;
  box-shadow: inset 0 1px 0 #80B0FB;
}

.Button {
  cursor: pointer;
  margin: 10px;
  padding: 10px;
  border-radius: 10px;
  font-size: 35px;
  font-weight: 800;
  color: white;
  background: #4C8FFB;
  border: 1px #3079ED solid;
  box-shadow: inset 0 1px 0 #80B0FB;
}

.Button:hover {
  border: 1px #2F5BB7 solid;
  box-shadow: 0 1px 1px #EAEAEA, inset 0 1px 0 #5A94F1;
  background: #3F83F1;
}

.Button:active {
  box-shadow: inset 0 2px 5px #2370FE;   
}