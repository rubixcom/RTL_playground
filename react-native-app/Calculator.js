import React, { Component } from 'react';
import { StyleSheet, Text, View, TouchableOpacity } from 'react-native';
//import { TouchableOpacity } from 'react-native-gesture-handler';

export default class Calculator extends Component {
  constructor(props) {
    super(props);
    this.state = {
      displayValue: '0',
      accumulator: 0,
      clearOnNext: true,
      operation: undefined
    }
  }

  performOperation() {
    let { accumulator, operation, displayValue } = this.state;
    if (operation === '+')
      accumulator += parseFloat(displayValue); 
    else if (operation === '-')
      accumulator -= parseFloat(displayValue); 
    else if (operation === '*')
      accumulator *= parseFloat(displayValue); 
    else if (operation === '/')
      accumulator /= parseFloat(displayValue); 
    return accumulator;
  }

  pressNum(value) {
    let { displayValue, clearOnNext } = this.state;
    if (clearOnNext) {
      displayValue = ''; clearOnNext = false;
    }
    if (displayValue !== '0' && value !== '0')
      displayValue += value;
    this.setState({ displayValue, clearOnNext });
  }

  pressDecimal() {
    let { displayValue, clearOnNext } = this.state;
    if (clearOnNext) {
      displayValue = '0'; clearOnNext = false;
    }
    if (displayValue.indexOf('.') === -1) {
      displayValue += '.';
      this.setState({ displayValue, clearOnNext });
    }
  }

  pressOperator(operation) {
    const { displayValue } = this.state;
    this.performOperation();
    const accumulator = parseFloat(displayValue);
    this.setState({ operation, accumulator, clearOnNext: true});
  }

  pressEquals() {
    const accumulator = this.performOperation();
    this.setState({ displayValue: accumulator.toString(), clearOnNext: true });
  }

  pressClear() {
    this.setState({ operation: undefined, displayValue: '0', clearOnNext: true, accumulator: 0 });
  }

  render() {
    const { displayValue } = this.state;
    return (
      <View style={styles.calculator}>
        <View style={styles.row}>
          <Text style={styles.displayPanel} testID="DisplayPanel">{displayValue}</Text>
        </View>
        <View style={styles.row}>
          <TouchableOpacity onPress={() => this.pressClear()} testID="C">
              <Text style={styles.button}>C</Text>
          </TouchableOpacity>
        </View>
        <View style={styles.row}>
          <TouchableOpacity onPress={() => this.pressNum('7')} testID="7">
              <Text style={styles.button}>7</Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={() => this.pressNum('8')} testID="8">
              <Text style={styles.button}>8</Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={() => this.pressNum('9')} testID="9">
              <Text style={styles.button}>9</Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={() => this.pressOperator('/')} testID="/">
              <Text style={styles.button}>รท</Text>
          </TouchableOpacity>
        </View>
        <View style={styles.row}>
          <TouchableOpacity onPress={() => this.pressNum('4')} testID="4">
              <Text style={styles.button}>4</Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={() => this.pressNum('5')} testID="5">
              <Text style={styles.button}>5</Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={() => this.pressNum('6')} testID="6">
              <Text style={styles.button}>6</Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={() => this.pressOperator('*')} testID="*">
              <Text style={styles.button}>*</Text>
          </TouchableOpacity>
        </View>
        <View style={styles.row}>
          <TouchableOpacity onPress={() => this.pressNum('1')} testID="1">
              <Text style={styles.button}>1</Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={() => this.pressNum('2')} testID="2">
              <Text style={styles.button}>2</Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={() => this.pressNum('3')}  testID="3">
              <Text style={styles.button}>3</Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={() => this.pressOperator('-')} testID="-">
              <Text style={styles.button}>-</Text>
          </TouchableOpacity>
        </View> 
        <View style={styles.row}>
          <TouchableOpacity onPress={() => this.pressNum('0')} testID="0">
              <Text style={styles.button}>0</Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={() => this.pressDecimal()} testID=".">
              <Text style={styles.button}>.</Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={() => this.pressEquals()} testID="=">
              <Text style={styles.button}>=</Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={() => this.pressOperator('+')} testID="+">
              <Text style={styles.button}>+</Text>
          </TouchableOpacity>
        </View> 
      </View>
    );
  }
}

const styles = StyleSheet.create({
  calculator: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'rgb(4, 28, 67)',
  },
  row: {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'flex-end',
    backgroundColor: 'rgb(4, 28, 67)',
    flexDirection: 'row',
    width: '100%',
  },
  displayPanel: {
    borderRadius: 10,
    fontSize: 50,
    color: '#fff',
    width: '94%',
    fontWeight: '800',
    textAlign: 'right',
    margin: 10,
    padding: 10,
    backgroundColor: 'rgb(4, 28, 67)',
    borderColor: '#3079ED',
    borderWidth: 3
  },
  button: {
    borderRadius: 10,
    color: '#fff',
    fontSize: 40,
    width: 80,
    height: 80,
    backgroundColor: '#4C8FFB',
    borderColor: '#3079ED',
    borderWidth: 3,
    fontWeight: '800',
    textAlign: 'center',
    justifyContent: 'center',
    alignItems: 'center',
    paddingTop: 10,
    margin: 10,
  },
});
