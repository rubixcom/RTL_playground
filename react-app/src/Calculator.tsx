import React from 'react';
import './Calculator.css';

enum Operator {
  multiply='ร', divide='รท', add='+', subtract='-'
}

interface Props { }
interface State {
  displayValue: string;
  accumulator: number;
  operation?: Operator;
  clearOnNext: boolean;
}

interface ButtonProps {
  testId: string,
  onClick: () => void,
  label: string
}

const Button = ({ testId, onClick, label }: ButtonProps) => (
  <div className="Button" data-testid={testId} onClick={onClick}>{label}</div>
)


class Calculator extends React.Component<Props, State> {
  constructor(props: React.Props<Props>) {
    super(props);
    this.state = {
      displayValue: '0',
      accumulator: 0,
      clearOnNext: true,
      operation: undefined
    }
  }

  performOperation() {
    let { accumulator, operation, displayValue } = this.state;
    if (operation === Operator.add)
      accumulator += parseFloat(displayValue); 
    else if (operation === Operator.subtract)
      accumulator -= parseFloat(displayValue); 
    else if (operation === Operator.multiply)
      accumulator *= parseFloat(displayValue); 
    else if (operation === Operator.divide)
      accumulator /= parseFloat(displayValue); 
    return accumulator;
  }

  pressNum(value: string) {
    let { displayValue, clearOnNext } = this.state;
    if (clearOnNext) {
      displayValue = ''; clearOnNext = false;
    }
    if (displayValue !== '0' && value !== '0')
      displayValue += value;
    this.setState({ displayValue, clearOnNext });
  }

  pressDecimal() {
    let { displayValue, clearOnNext } = this.state;
    if (clearOnNext) {
      displayValue = '0'; clearOnNext = false;
    }
    if (displayValue.indexOf('.') === -1) {
      displayValue += '.';
      this.setState({ displayValue, clearOnNext });
    }
  }

  pressOperator(operation: Operator) {
    const { displayValue } = this.state;
    this.performOperation();
    const accumulator = parseFloat(displayValue);
    this.setState({ operation, accumulator, clearOnNext: true});
  }

  pressEquals() {
    const accumulator = this.performOperation();
    this.setState({ displayValue: accumulator.toString(), clearOnNext: true });
  }

  pressClear() {
    this.setState({ operation: undefined, displayValue: '0', clearOnNext: true, accumulator: 0 });
  }

  render() {
    const { displayValue } = this.state;
    return (
      <div className="App">
        <div className="Calculator">
          <div className="DisplayPanel" data-testid="DisplayPanel">{displayValue}</div>
          <div></div>
          <div></div>
          <div></div>
          <Button onClick={() => this.pressClear()} testId={'C'} label={'C'}></Button>
          <Button onClick={() => this.pressNum('7')} testId={'7'} label={'7'}></Button>
          <Button onClick={() => this.pressNum('8')} testId={'8'} label={'8'}></Button>
          <Button onClick={() => this.pressNum('9')} testId={'9'} label={'9'}></Button>
          <Button onClick={() => this.pressOperator(Operator.divide)} testId={'/'} label={'รท'}></Button>
          <Button onClick={() => this.pressNum('4')} testId={'4'} label={'4'}></Button>
          <Button onClick={() => this.pressNum('5')} testId={'5'} label={'5'}></Button>
          <Button onClick={() => this.pressNum('6')} testId={'6'} label={'6'}></Button>
          <Button onClick={() => this.pressOperator(Operator.multiply)} testId={'*'} label={'x'}></Button>
          <Button onClick={() => this.pressNum('1')} testId={'1'} label={'1'}></Button>
          <Button onClick={() => this.pressNum('2')} testId={'2'} label={'2'}></Button>
          <Button onClick={() => this.pressNum('3')} testId={'3'} label={'3'}></Button>
          <Button onClick={() => this.pressOperator(Operator.subtract)} testId={'-'} label={'-'}></Button>
          <Button onClick={() => this.pressNum('0')} testId={'0'} label={'0'}></Button>
          <Button onClick={() => this.pressDecimal()} testId={'.'} label={'.'}></Button>
          <Button onClick={() => this.pressEquals()} testId={'='} label={'='}></Button>
          <Button onClick={() => this.pressOperator(Operator.add)} testId={'+'} label={'+'}></Button>
        </div>
      </div>
    );
  }
}

export default Calculator;
