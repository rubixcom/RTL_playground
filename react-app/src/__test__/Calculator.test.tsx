import React from 'react';
import { render, cleanup, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import Calculator from '../Calculator';

afterEach(cleanup);

it("matches snapshot", () => {
  const { asFragment } = render(<Calculator />);
  expect(asFragment()).toMatchSnapshot();
});

it('displays 0 when first initialised', () => {
    const { getByTestId } = render(<Calculator />);
    const displayPanel = getByTestId('DisplayPanel');
    expect(displayPanel.textContent).toBe('0');  
});

it('supports basic entry of values', () => {
    const { getByTestId, getByText } = render(<Calculator />);
    fireEvent.click(getByText('1'));
    fireEvent.click(getByText('.'));
    fireEvent.click(getByText('2'));
    fireEvent.click(getByText('3'));
    const displayPanel = getByTestId('DisplayPanel');
    expect(displayPanel.textContent).toBe('1.23');  
});

it('doesn\'t accept invalid numbers', () => {
    const { getByTestId, getByText } = render(<Calculator />);
    fireEvent.click(getByText('1'));
    fireEvent.click(getByText('.'));
    fireEvent.click(getByText('2'));
    fireEvent.click(getByText('.'));
    fireEvent.click(getByText('3'));
    const displayPanel = getByTestId('DisplayPanel');
    expect(displayPanel.textContent).toBe('1.23');  
});

it('does not display leading 0', () => {
    const { getByTestId, getByText } = render(<Calculator />);
    fireEvent.click(getByText('.'));
    fireEvent.click(getByText('2'));
    fireEvent.click(getByText('3'));
    const displayPanel = getByTestId('DisplayPanel');
    expect(displayPanel.textContent).toBe('0.23');  
});

it('doesn\'t allow allow multiple 0s', () => {
    const { getByTestId } = render(<Calculator />);
    fireEvent.click(getByTestId('0'));
    fireEvent.click(getByTestId('0'));
    fireEvent.click(getByTestId('3'));
    const displayPanel = getByTestId('DisplayPanel');
    expect(displayPanel.textContent).toBe('3');  
});

it('does addition', () => {
    const { getByTestId } = render(<Calculator />);
    const displayPanel = getByTestId('DisplayPanel');
    fireEvent.click(getByTestId('2'));
    expect(displayPanel.textContent).toBe('2');  
    fireEvent.click(getByTestId('+'));
    fireEvent.click(getByTestId('2'));
    expect(displayPanel.textContent).toBe('2');  
    fireEvent.click(getByTestId('='));
    expect(displayPanel.textContent).toBe('4');  
});

it('it does repeat operations', () => {
    const { getByTestId } = render(<Calculator />);
    fireEvent.click(getByTestId('2'));
    fireEvent.click(getByTestId('*'));
    fireEvent.click(getByTestId('2'));
    fireEvent.click(getByTestId('='));
    fireEvent.click(getByTestId('='));
    fireEvent.click(getByTestId('='));
    fireEvent.click(getByTestId('='));
    const displayPanel = getByTestId('DisplayPanel');
    expect(displayPanel.textContent).toBe('32');  
});

it('does long series or operations', () => {
    const { getByTestId } = render(<Calculator />);
    const displayPanel = getByTestId('DisplayPanel');
    fireEvent.click(getByTestId('9'));
    fireEvent.click(getByTestId('8'));
    fireEvent.click(getByTestId('7'));
    expect(displayPanel.textContent).toBe('987');  
    fireEvent.click(getByTestId('/'));
    fireEvent.click(getByTestId('4'));
    expect(displayPanel.textContent).toBe('4');  
    fireEvent.click(getByTestId('*'));
    fireEvent.click(getByTestId('5'));
    fireEvent.click(getByTestId('6'));
    fireEvent.click(getByTestId('-'));
    fireEvent.click(getByTestId('1'));
    fireEvent.click(getByTestId('2'));
    fireEvent.click(getByTestId('3'));
    fireEvent.click(getByTestId('4'));
    expect(displayPanel.textContent).toBe('1234');  
    fireEvent.click(getByTestId('='));
    expect(displayPanel.textContent).toBe('-1178');  
    fireEvent.click(getByTestId('C'));
    expect(displayPanel.textContent).toBe('0');  
    fireEvent.click(getByTestId('1'));
    fireEvent.click(getByTestId('2'));
    fireEvent.click(getByTestId('3'));
    expect(displayPanel.textContent).toBe('123');  
    fireEvent.click(getByTestId('='));
    expect(displayPanel.textContent).toBe('0');  
});