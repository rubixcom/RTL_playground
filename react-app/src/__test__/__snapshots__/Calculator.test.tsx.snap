// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="Calculator"
    >
      <div
        class="DisplayPanel"
        data-testid="DisplayPanel"
      >
        0
      </div>
      <div />
      <div />
      <div />
      <div
        class="Button"
        data-testid="C"
      >
        C
      </div>
      <div
        class="Button"
        data-testid="7"
      >
        7
      </div>
      <div
        class="Button"
        data-testid="8"
      >
        8
      </div>
      <div
        class="Button"
        data-testid="9"
      >
        9
      </div>
      <div
        class="Button"
        data-testid="/"
      >
        รท
      </div>
      <div
        class="Button"
        data-testid="4"
      >
        4
      </div>
      <div
        class="Button"
        data-testid="5"
      >
        5
      </div>
      <div
        class="Button"
        data-testid="6"
      >
        6
      </div>
      <div
        class="Button"
        data-testid="*"
      >
        x
      </div>
      <div
        class="Button"
        data-testid="1"
      >
        1
      </div>
      <div
        class="Button"
        data-testid="2"
      >
        2
      </div>
      <div
        class="Button"
        data-testid="3"
      >
        3
      </div>
      <div
        class="Button"
        data-testid="-"
      >
        -
      </div>
      <div
        class="Button"
        data-testid="0"
      >
        0
      </div>
      <div
        class="Button"
        data-testid="."
      >
        .
      </div>
      <div
        class="Button"
        data-testid="="
      >
        =
      </div>
      <div
        class="Button"
        data-testid="+"
      >
        +
      </div>
    </div>
  </div>
</DocumentFragment>
`;
